File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[]}
     */
    productExceptSelf(nums) {
        const n = nums.length;
        let product = 1;
        let zeroCount = 0;
        for(let num of nums){
            if(num === 0) zeroCount++;
            else product *= num;
        }

        let ans = new Array(n).fill(0);

        if(zeroCount > 1) return ans;

        for(let i = 0; i < n; i++) {
            if(zeroCount > 0) {
                ans[i] = (nums[i] === 0 ? product : 0);
            } else {
                ans[i] = product / nums[i];
            }
        }

        return ans;
    }
}

File: Data Structures & Algorithms/top-k-elements-in-list/submission-0.cpp
class Solution {
    public:
        vector<int> topKFrequent(vector<int>& nums, int k) {
            unordered_map<int, int> freq;
            for(int num : nums){
                freq[num]++;
            }

            int n = nums.size();
            vector<vector<int>> buckets(n + 1);
            for(auto it : freq){
                int num = it.first;
                int freq = it.second;
                buckets[freq].push_back(num);
            }

            vector<int> ans;
            for(int i = buckets.size() - 1; i >= 0; i--){
                for(int num : buckets[i]){
                    if(ans.size() == k) break;
                    ans.push_back(num);
                }
            }

            return ans;
        }
};